### DISCUSSION TITLE
Instance (own) properties vs. `prototype` properties.

### DISCUSSION THREAD

**Discussion by ghost** [mdn/mdn-community#153] [Content]
To start us off, I'll reference some issues that have already arisen regarding this topic:
https://github.com/mdn/content/issues/16325
https://github.com/mdn/content/issues/17610
https://github.com/mdn/content/issues/18188
https://github.com/mdn/content/discussions/5121

The problem that some readers have with JS documentation is that it currently reads as if to claim that instance properties are instead located on the prototype - which is misleading, but not clarified either.

At the most basic level, this is simply incorrect. MDN makes references to "own" properties, we have articles about prototypes and inheritance, and there are methods such as `Object.hasOwn`, so if these are shown as on a prototype, but they aren't, then it may very well lead to confusion, thus leading to more issues about this from readers.

For example, in the issue that I raised, [`Error`](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Error) claims that [`cause`](https://developer.mozilla.org/docs/Web/JavaScript/Reference/Global_Objects/Error/cause) is on the prototype, yet it's not even necessarily on the instance.

The most accurate way to describe it would be "optionally present on the instance," because if `{ cause }` is not present in the constructor, then the property isn't even installed onto the object.
```js
assert(
 Object.hasOwn(new Error("", { cause: null }), "cause")
);
assert(
 !Object.hasOwn(new Error(""), "cause")
);
assert(
 !Object.hasOwn(Error.prototype, "cause")
);
```

This issue applies to `Function.{name, length}`, `Error.{message, cause}`, and a few others.

It seems that the current resolution is to always use `.prototype.`, see: https://github.com/mdn/content/issues/16325#issuecomment-1135302663.

May this be reconsidered?

**Comment by teoli2003** [comment]
The current resolution came from the understanding that there are 2 types of properties (static and on the prototype.

It looks like we have 3 types of properties?
- static
- "own" properties on the prototype
- "own" properties not coming from the prototype

Am I right? (Likely not)

**Reply by Josh-Cena** [reply]
> "own" properties on the prototype

Only properties _not_ on the prototype (i.e. on the object itself) are "own". But otherwise, yes, that's basically the idea.

**Reply by teoli2003** [reply]
So:
- static properties
- prototype properties
- own properties (not on the prototype, not static)

Is there the same triad for methods?

**Reply by Josh-Cena** [reply]
Methods are hardly every defined on the instance, because they are polymorphic by `this` so they only need to be defined on the prototype. I'm only aware of own non-function properties.

**Reply by ghost** [reply]
Do note that "own" property is also relative. A "static" property is a constructor's own property, and a prototype property is a prototype's own property.

**Reply by ghost** [reply]
> So:
> - static properties
> - prototype properties
> - own properties (not on the prototype, not static)
> 
> Is there the same triad for methods?

Yup! Methods are merely properties with functions as values. (Well, specifically accepting a `this` parameter)

**Reply by js-choi** [reply]
I agree that this is an important distinction that actually affects developers’ experience.

I think the way that the Ecma262 specification itself handles the naming is instructive. It tends to use the phrase “Properties of X Instances” (e.g., [“Properties of Array Instances”](https://tc39.es/ecma262/#sec-properties-of-array-instances)) as opposed to “Properties of the X Prototype” (e.g., [“Properties of the Array Prototype Object”](https://tc39.es/ecma262/#sec-properties-of-the-array-prototype-object)) and “Properties of the X Constructor” (e.g., [“Properties of the Array Constructor”](https://tc39.es/ecma262/#sec-properties-of-the-array-constructor)).

So, we could retitle the “Array.prototype.length” page as “Array instance length”, and it would be still concise and clear enough.

**Reply by teoli2003** [reply]
cc/ @Rumyra who is leading the page renaming project.

**Reply by Rumyra** [reply]
Thanks @teoli2003 :)

I'm happy to consider the renaming of JS pages much like we discussed the Web API pages, but before we take this further I would like us to have done some Web API renaming first. Reasons:

* It'd be good to understand the effort/time involved which we can judge with Web API renaming
* I'm interested in the feedback from users (if any) after renaming a few Web API pages. What happens there might influence our decisions for this discussion

**Reply by ghost** [reply]
> ... but before we take this further I would like us to have done some Web API renaming first. ...

I'm not sure what you mean, do you refer to Web API renaming as in renaming some Web API pages to conform to this discussion's topic about having instance properties not on the prototype?
If so, I don't believe that this discussion applies to any Web APIs though, rather only ECMAScript objects.

**Reply by js-choi** [reply]
I think what @Rumyra is saying—please correct me if I’m wrong—is that they think that it might be worth seeing if renaming Web API pages to more appropriate names will get a lot of positive negative feedback from developers, since renaming Ecma262 API pages might garner similar community feedback. 

I’m not sure what sort of feedback would necessarily affect this decision—incorrect is incorrect, and titles like “Function.length” or “Array.prototype.length” for function/array lengths in general are observably incorrect and definitely misleading...but it’s fine by me to be cautious and wait to see what happens with a similar situation.

**Reply by ghost** [reply]
Yeah, but but it seems like this,
> renaming Web API pages to more appropriate names

is a reference to a separate issue. If so, and anyone knows where that is, might someone point me to it? Otherwise, again, I don't understand any connection/relevance.

Although, is anything being done for accessors? That would make sense, and apply to Web APIs, if that's what they were referring to? ...

**Reply by Josh-Cena** [reply]
@Phosra I think it's referring to https://github.com/mdn/content/discussions/5121 although I don't have much idea how that's progressing

**Reply by ghost** [reply]
Ah, understood, I had thought that it might have already been finished.
